cpp/src/arrow/dataset/file_skyhook.h

```diff
@@ -0,0 +1,275 @@
+
+  bool splittable() const { return true; }
+
+  bool Equals(const FileFormat& other) const override {
+    return type_name() == other.type_name();
+  }
+
+  Result<bool> IsSupported(const FileSource& source) const override { return true; }
+
+  /// \brief Return the schema of the file fragment.
+  /// \param[in] source The source of the file fragment.
+  /// \return The schema of the file fragment.
+  Result<std::shared_ptr<Schema>> Inspect(const FileSource& source) const override;
+
+  /// \brief Scan a file fragment.
+  /// \param[in] options Options to pass.
+  /// \param[in] file The file fragment.
+  /// \return The scanned file fragment.
+  Result<ScanTaskIterator> ScanFile(
+      const std::shared_ptr<ScanOptions>& options,
+      const std::shared_ptr<FileFragment>& file) const override;
+
+  Result<std::shared_ptr<FileWriter>> MakeWriter(
+      std::shared_ptr<io::OutputStream> destination, std::shared_ptr<Schema> schema,
+      std::shared_ptr<FileWriteOptions> options) const {
+    return Status::NotImplemented("Use the Python API");
+  }
+
+  std::shared_ptr<FileWriteOptions> DefaultWriteOptions() override { return NULLPTR; }
+
+ protected:
+  std::shared_ptr<SkyhookDirectObjectAccess> doa_;
+  std::string fragment_format_;
+};
+
+/// \brief Serialize scan request to a bufferlist.
+/// \param[in] options The scan options to use to build a ScanRequest.
+/// \param[in] file_format The underlying file format to use.
+/// \param[in] file_size The size of the file fragment.
+/// \param[out] bl Output bufferlist.
+/// \return Status.
+ARROW_DS_EXPORT Status SerializeScanRequest(std::shared_ptr<ScanOptions>& options,
+                                            int& file_format, int64_t& file_size,
+                                            ceph::bufferlist& bl);
+
+/// \brief Deserialize scan request from bufferlist.
+/// \param[out] filter The filter expression to apply.
+/// \param[out] partition The partition expression to use.
+/// \param[out] projected_schema The schema to project the filtered record batches.
+/// \param[out] dataset_schema The dataset schema to use.
+/// \param[out] file_size The size of the file fragment.
+/// \param[out] file_format The underlying file format to use.
+/// \param[in] bl Input Ceph bufferlist.
+/// \return Status.
+ARROW_DS_EXPORT Status DeserializeScanRequest(compute::Expression* filter,
+                                              compute::Expression* partition,
+                                              std::shared_ptr<Schema>* projected_schema,
+                                              std::shared_ptr<Schema>* dataset_schema,
+                                              int64_t& file_size, int& file_format,
+                                              ceph::bufferlist& bl);
+
+/// \brief Serialize the result Table to a bufferlist.
+/// \param[in] table The table to serialize.
+/// \param[in] aggressive If true, use ZSTD compression instead of LZ4.
```

Comment:
Is there a reason to hardcode ZSTD/LZ4 instead of accepting an Arrow codec instance or codec type?